tsdb/head_read.go

```diff
@@ -646,8 +652,13 @@ func (s *memSeries) iterator(id chunks.HeadChunkID, isoState *isolationState, ch
 			}
 		}
 
-		if s.headChunk != nil {
-			totalSamples += s.headChunk.chunk.NumSamples()
+		ix -= len(s.mmappedChunks)
+
+		for j, d := range s.headChunks {
+			totalSamples += d.chunk.NumSamples()
+			if j < ix {
+				previousSamples += d.chunk.NumSamples()
+			}
```

Comment:
We can call `d.chunk.NumSamples()` only once by assigning it to a variable. It actually reads the chunk bytes and decodes the number.